File: src/services/checkin.ts
/**
 * 签到服务
 */

import type { Provider, CheckinResponse } from '@/types/provider';

/**
 * 检查供应商是否支持签到
 */
export function isCheckinSupported(provider: Provider): boolean {
  // 必须有 API 地址和认证信息
  if (!provider.api_address && !provider.website_url) return false;
  if (!provider.user_id || !provider.access_token) return false;
  return true;
}

/**
 * 执行签到
 */
export async function performCheckin(provider: Provider): Promise<CheckinResponse> {
  if (!isCheckinSupported(provider)) {
    return {
      success: false,
      message: '供应商缺少必要的签到配置信息',
    };
  }

  const endpoint = provider.checkin_config?.endpoint || '/api/user/checkin';
  const baseUrl = provider.api_address || provider.website_url;
  const url = `${baseUrl.replace(/\/$/, '')}${endpoint}`;

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${provider.access_token}`,
        'New-Api-User': provider.user_id,
        'Content-Type': 'application/json',
      },
    });

    // 检查 HTTP 状态码
    if (!response.ok) {
      // 404 可能表示不支持签到功能
      if (response.status === 404) {
        return {
          success: false,
          message: '该供应商不支持签到功能 (404)',
        };
      }
      
      // 尝试解析错误响应
      try {
        const errorData = await response.json();
        return {
          success: false,
          message: errorData.message || `请求失败 (${response.status})`,
        };
      } catch {
        return {
          success: false,
          message: `请求失败 (${response.status})`,
        };
      }
    }

    const data = await response.json();
    
    // 检查响应格式
    if (typeof data.success !== 'boolean') {
      return {
        success: false,
        message: '供应商返回的数据格式不正确',
      };
    }

    return data as CheckinResponse;
  } catch (error) {
    // 网络错误或其他异常
    if (error instanceof TypeError && error.message.includes('fetch')) {
      return {
        success: false,
        message: '网络连接失败，请检查供应商地址是否正确',
      };
    }
    
    return {
      success: false,
      message: error instanceof Error ? error.message : '签到请求失败',
    };
  }
}

/**
 * 获取签到状态
 */
export async function getCheckinStatus(provider: Provider): Promise<CheckinResponse> {
  if (!isCheckinSupported(provider)) {
    return {
      success: false,
      message: '供应商缺少必要的签到配置信息',
    };
  }

  const endpoint = provider.checkin_config?.endpoint || '/api/user/checkin';
  const baseUrl = provider.api_address || provider.website_url;
  const url = `${baseUrl.replace(/\/$/, '')}${endpoint}`;

  try {
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${provider.access_token}`,
        'New-Api-User': provider.user_id,
      },
    });

    // 检查 HTTP 状态码
    if (!response.ok) {
      // 404 可能表示不支持签到功能
      if (response.status === 404) {
        return {
          success: false,
          message: '该供应商不支持签到功能',
        };
      }
      
      // 尝试解析错误响应
      try {
        const errorData = await response.json();
        return {
          success: false,
          message: errorData.message || `请求失败 (${response.status})`,
        };
      } catch {
        return {
          success: false,
          message: `请求失败 (${response.status})`,
        };
      }
    }

    const data = await response.json();
    
    // 检查响应格式
    if (typeof data.success !== 'boolean') {
      return {
        success: false,
        message: '供应商返回的数据格式不正确',
      };
    }

    return data as CheckinResponse;
  } catch (error) {
    // 网络错误或其他异常
    if (error instanceof TypeError && error.message.includes('fetch')) {
      return {
        success: false,
        message: '网络连接失败，请检查供应商地址是否正确',
      };
    }
    
    return {
      success: false,
      message: error instanceof Error ? error.message : '获取签到状态失败',
    };
  }
}

File: src-tauri/src/services/checkin.rs
// Checkin Service
//
// 供应商签到服务

use crate::models::provider::{CheckinConfig, Provider};
use anyhow::{anyhow, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::time::Duration;

#[derive(Debug, Serialize, Deserialize)]
pub struct CheckinResponse {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<CheckinData>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CheckinData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quota_awarded: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub checkin_date: Option<String>,
}

/// 执行签到
pub async fn perform_checkin(provider: &Provider) -> Result<CheckinResponse> {
    let config = provider
        .checkin_config
        .as_ref()
        .ok_or_else(|| anyhow!("签到配置不存在"))?;

    let base_url = provider
        .api_address
        .as_ref()
        .unwrap_or(&provider.website_url);
    let url = format!(
        "{}{}",
        base_url.trim_end_matches('/'),
        config.endpoint
    );

    let client = Client::builder()
        .timeout(Duration::from_secs(30))
        .build()?;

    let response = client
        .post(&url)
        .header("Authorization", format!("Bearer {}", provider.access_token))
        .header("New-Api-User", &provider.user_id)
        .header("Content-Type", "application/json")
        .send()
        .await?;

    if !response.status().is_success() {
        let status = response.status();
        let error_text = response.text().await.unwrap_or_default();
        return Err(anyhow!("签到失败 ({}): {}", status, error_text));
    }

    let result: CheckinResponse = response.json().await?;
    Ok(result)
}

/// 检查是否需要签到
pub fn should_checkin(config: &CheckinConfig) -> bool {
    if !config.enabled {
        return false;
    }

    // 检查今天是否已签到
    if let Some(last_checkin) = config.last_checkin_at {
        let now = chrono::Local::now();
        let last = chrono::DateTime::<chrono::Local>::from(
            chrono::DateTime::<chrono::Utc>::from_timestamp(last_checkin, 0)
                .unwrap_or_default(),
        );

        // 如果是同一天,不需要签到
        if now.date_naive() == last.date_naive() {
            return false;
        }
    }

    // 检查当前时间是否到达签到时间
    let now = chrono::Local::now();
    let current_hour = now.hour() as u8;

    // 如果当前时间大于等于设定的签到时间,则需要签到
    current_hour >= config.checkin_hour
}
